src/window/mod.rs

```diff
@@ -0,0 +1,601 @@
+//! [`StatefulWindowUnary::stateful_window_unary`].
+//!
+//! All user-facing windowing operators like
+//! [`crate::dataflow::Dataflow::reduce_window`] are built on top of
+//! [`StatefulWindowUnary::stateful_window_unary`] so we only have to
+//! write the window management code once.
+//!
+//! That operator itself is based upon
+//! [`crate::operators::StatefulUnary`], which provides a generic
+//! abstraction to the recovery system. We get recovery for "free" as
+//! long as we play by the rules of that operator.
+//!
+//! This system follows our standard pattern of having parallel Python
+//! config objects and Rust impl structs for each trait of behavior we
+//! want. E.g. [`SystemClockConfig`] represents a token in Python for
+//! how to create a [`SystemClock`].
+
+use crate::dataflow::StepId;
+use crate::operators::StatefulUnary;
+use crate::pyo3_extensions::{StateKey, TdPyAny, TdPyCallable};
+use crate::recovery::StateBackup;
+use crate::with_traceback;
+use chrono::{Duration, NaiveDateTime};
+use log::debug;
+use pyo3::exceptions::PyValueError;
+use pyo3::prelude::*;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::task::Poll;
+use timely::dataflow::{Scope, Stream};
+use timely::{Data, ExchangeData};
+
+/// Base class for a clock config.
+///
+/// This describes how a windowing operator should determine the
+/// current time and the time for each element.
+///
+/// Use a specific subclass of this that matches the time definition
+/// you'd like to use.
+#[pyclass(module = "bytewax.window", subclass)]
+#[pyo3(text_signature = "()")]
+pub(crate) struct ClockConfig;
+
+/// Use the system time inside the windowing operator to determine
+/// times.
+///
+/// If the dataflow has no more input, all windows are closed.
+///
+/// Returns:
+///
+///     Config object. Pass this as the `clock_config` parameter to
+///     your windowing operator.
+#[pyclass(module="bytewax.window", extends=ClockConfig)]
+struct SystemClockConfig {}
+
+#[pymethods]
+impl SystemClockConfig {
+    #[new]
+    #[args()]
+    fn new() -> (Self, ClockConfig) {
+        (Self {}, ClockConfig {})
+    }
+
+    /// Pickle as a tuple.
+    fn __getstate__(&self) -> (&str,) {
+        ("SystemClockConfig",)
+    }
+
+    /// Unpickle from tuple of arguments.
+    fn __setstate__(&mut self, state: &PyAny) -> PyResult<()> {
+        if let Ok(("SystemClockConfig",)) = state.extract() {
+            Ok(())
+        } else {
+            Err(PyValueError::new_err(format!(
+                "bad pickle contents for SystemClockConfig: {state:?}"
+            )))
+        }
+    }
+}
+
+pub(crate) fn build_clock<V>(
+    py: Python,
+    clock_config: Py<ClockConfig>,
+) -> Result<Box<dyn Clock<V>>, String> {
+    let window_config = clock_config.as_ref(py);
+
+    if let Ok(system_clock_config) = window_config.downcast::<PyCell<SystemClockConfig>>() {
+        let _system_clock_config = system_clock_config.borrow();
```

Comment:
If we ever needed to pull config parameters out of the config object, yes. But there are no configuration parameters currently, so kinda no. I wanted to leave this no-op in there so it's a reminder that you need to do this for new config types. See https://github.com/bytewax/bytewax/blob/772c51f7c7cca09321aa86cecb34768148bd6b13/src/recovery/mod.rs#L672-L735 for the pattern I'm copying. It just looks goofy here because we don't have any other config types.